src/main/java/io/github/classgraph/ClasspathElementDir.java

```diff
@@ -396,16 +401,19 @@ private void scanPathRecursively(final Path path, final LogNode log) {
             return;
         }
         Collections.sort(pathsInDir);
+        FileUtils.FileAttributesGetter getFileAttributes = FileUtils.createCachedAttributesGetter();
 
         // Determine whether this is a modular jar running under JRE 9+
         final boolean isModularJar = VersionFinder.JAVA_MAJOR_VERSION >= 9 && getModuleName() != null;
 
         // Only scan files in directory if directory is not only an ancestor of an accepted path
         if (parentMatchStatus != ScanSpecPathMatch.ANCESTOR_OF_ACCEPTED_PATH) {
             // Do preorder traversal (files in dir, then subdirs), to reduce filesystem cache misses
-            for (final Path subPath : pathsInDir) {
+            for (final Path subPath : new ArrayList<>(pathsInDir)) {
```

Comment:
I think you could avoid the copy of the list and the subsequent linear scans for `pathsInDir.remove` by using an Iterator instead of the for-loop along with `Iterator.remove`.

Reply:
Indeed, I just selected one of the smallest & easy change to avoid any side effects and considered that one also rather cheap, at least negligible regarding the whole performance.